Update CI.yml

PhiSpel opened the pull request:
try running CI on test-branch. I have issues with some import statements at https://github.com/PhiSpel/lettuce/tree/cuda-native-philipp

Olllom commented on lettuce/flows/decayingturbulence.py:
```diff
@@ -120,7 +106,6 @@ def initial_solution(self, x):
         u = np.append(u, u1f.real[None, ...], axis=0)
 
         p = (u[0]*0)[None,...]
-        p = self.pressure_poisson(p,u)
```
This should now be done manually by calling
`simulation.initialize_pressure`

Olllom commented on lettuce/flows/decayingturbulence.py:
```diff
@@ -145,18 +131,6 @@ def __init__(self, resolution, reynolds_number, mach_number, lattice, k0=20, ic_
             characteristic_velocity_pu=1
         )
 
-    def pressure_poisson(self, p, u ):
-        dx = self.units.characteristic_length_pu / self.resolution
-        u = torch.tensor(u, device=self.units.lattice.device, dtype=self.units.lattice.dtype)
-
-        div_u0 = torch_gradient(torch_gradient(u[0] * u[0], dx)[0] + torch_gradient(u[0] * u[1], dx)[1] + torch_gradient(u[0] * u[2], dx)[2], dx)[0]  # uii+uij
-        div_u1 = torch_gradient(torch_gradient(u[1] * u[0], dx)[0] + torch_gradient(u[1] * u[1], dx)[1] + torch_gradient(u[1] * u[2], dx)[2], dx)[1]  # uji+ujj
-        div_u2 = torch_gradient(torch_gradient(u[2] * u[0], dx)[0] + torch_gradient(u[2] * u[1], dx)[1] + torch_gradient(u[2] * u[2], dx)[2], dx)[2]  # uji+ujj
-        u_mod = -(div_u0 + div_u1 + div_u2)
-
-        p_mod = torch_jacobi(u_mod, p[0], dx, self.units.lattice.device,dim=3, tol_abs=1e-6)[None, ...]
-        return p_mod
-
```
same as above

Olllom commented on lettuce/simulation.py:
```diff
@@ -1,13 +1,17 @@
 
 from timeit import default_timer as timer
-from lettuce import LettuceException, get_default_moment_transform, BGKInitialization, ExperimentalWarning, torch_gradient
+from lettuce import (
+    LettuceException, get_default_moment_transform, BGKInitialization, ExperimentalWarning, torch_gradient
+)
+from lettuce.util import pressure_poisson
```
just cosmetic

Olllom commented on lettuce/simulation.py:
```diff
@@ -80,6 +84,20 @@ def initialize(self, max_num_steps=500, tol_pressure=0.001):
             p_old = deepcopy(p)
         return i
 
+    def initialize_pressure(self, max_num_steps=100000, tol_pressure=1e-6):
+        """Reinitialize equilibrium distributions with pressure obtained by a Jacobi solver.
+        Note that this method has to be called before initialize_f_neq.
+        """
+        u = self.lattice.u(self.f)
+        rho = pressure_poisson(
+            self.flow.units,
+            self.lattice.u(self.f),
+            self.lattice.rho(self.f),
+            tol_abs=tol_pressure,
+            max_num_steps=max_num_steps
+        )
+        self.f = self.lattice.equilibrium(rho, u)
+
```
this calls the jacobi stuff and is available for all flows

Olllom commented on lettuce/simulation.py:
```diff
@@ -91,19 +109,21 @@ def initialize_f_neq(self):
         grad_u1 = torch_gradient(u[1], dx=1, order=6)[None, ...]
         S = torch.cat([grad_u1, grad_u0])
 
-        if(self.lattice.D==3):
+        if self.lattice.D == 3:
             grad_u2 = torch_gradient(u[2], dx=1, order=6)[None, ...]
             S = torch.cat([S, grad_u2])
 
-        Pi_1 = -1.0 * self.flow.units.relaxation_parameter_lu * rho  * S / self.lattice.cs**2
-        Q = torch.einsum('ia,ib->iab',self.lattice.e,self.lattice.e)-torch.eye(self.lattice.D,device=self.lattice.device)*self.lattice.cs**2
+        Pi_1 = -1.0 * self.flow.units.relaxation_parameter_lu * rho * S / self.lattice.cs**2
+        Q = (
+            torch.einsum('ia,ib->iab', self.lattice.e,self.lattice.e)
+            - torch.eye(self.lattice.D, device=self.lattice.device)*self.lattice.cs**2
+        )
```
cosmetic

Olllom commented on lettuce/util.py:
```diff
@@ -90,7 +90,8 @@ def torch_gradient(f, dx=1, order=2):
     return out
 
-def grid_fine_to_coarse(lattice,f_fine,tau_fine,tau_coarse):
+
+def grid_fine_to_coarse(lattice, f_fine, tau_fine, tau_coarse):
```
cosmetic

Olllom commented on lettuce/util.py:
```diff
@@ -100,14 +101,17 @@ def grid_fine_to_coarse(lattice,f_fine,tau_fine,tau_coarse):
-    p = torch.tensor(p, device=device, dtype=torch.double)
-    dx = torch.tensor(dx, device=device, dtype=torch.double)
+
+def torch_jacobi(f, p, dx, device, dim, tol_abs=1e-10, max_num_steps=100000):
```
added the num steps as a parameter

Olllom commented on lettuce/util.py:
```diff
@@ -134,7 +138,53 @@ def torch_jacobi(f, p, dx, device, dim, tol_abs=1e-10):
+    return p
+
+
+def pressure_poisson(units, u, rho0, tol_abs=1e-10, max_num_steps=100000):
```
This is the new stuff (moved from decaying)

Olllom commented on lettuce/util.py:
```diff
@@ -134,7 +138,53 @@ def torch_jacobi(f, p, dx, device, dim, tol_abs=1e-10):
+        The flow instance.
+    u : torch.Tensor
+        The velocity tensor.
+    rho0 : torch.Tensor
```
I use rho instead of p, because we are now doing this in torch (no numpy!)
The philosophy is that everything in the flow classes is in numpy and physical units. 
Everything in the simulation is in torch and lattice units.
That's why I am expecting lattice units input and output in lattice units as well.

Olllom commented on tests/test_flows.py:
```diff
@@ -6,7 +6,7 @@
 from lettuce import DecayingTurbulence2D, DecayingTurbulence3D, torch_gradient
```
all changes cosmetic in this file

Olllom commented on tests/test_simulation.py:
```diff
@@ -22,11 +22,11 @@ def test_save_and_load(dtype_device, tmpdir):
 
-
-def test_initialization(dtype_device):
+@pytest.mark.parametrize("use_jacobi", [True, False])
```
this is now testing the API with pressure initialization (it works!)

Olllom commented on tests/test_simulation.py:
```diff
@@ -59,10 +64,11 @@ def test_initialize_fneq(Case, dtype_device):
-    assert(torch.allclose(pre_u,post_u))
+    tol = 1e-6 if dtype is torch.float32 else 1e-10
+    assert(torch.allclose(pre_rho, post_rho, rtol=0.0, atol=tol))
+    assert(torch.allclose(pre_u, post_u, rtol=0.0, atol=tol))
```
this test was failing (no wonder, using relative tolerance is not a good idea)

Olllom commented on tests/test_util.py:
```diff
@@ -36,17 +40,18 @@ def test_torch_gradient_3d(order):
 
-def test_grid_fine_to_couse_2d():
+
+def test_grid_fine_to_coarse_2d():
```
I sped this up. This took ages in the comparison.

Olllom commented on tests/test_util.py:
```diff
@@ -56,32 +61,53 @@ def test_grid_fine_to_couse_2d():
+
+
+@pytest.mark.parametrize("Stencil", [D2Q9, D3Q27])
+def test_pressure_poisson(dtype_device, Stencil):
+    dtype, device = dtype_device
+    lattice = Lattice(Stencil(), device, dtype)
+    flow_class = TaylorGreenVortex2D if Stencil is D2Q9 else TaylorGreenVortex3D
+    flow = flow_class(resolution=32, reynolds_number=100, mach_number=0.05, lattice=lattice)
+    p0, u = flow.initial_solution(flow.grid)
+    u = flow.units.convert_velocity_to_lu(lattice.convert_to_tensor(u))
+    rho0 = flow.units.convert_pressure_pu_to_density_lu(lattice.convert_to_tensor(p0))
+    rho = pressure_poisson(flow.units, u, torch.ones_like(rho0))
+    pfinal = flow.units.convert_density_lu_to_pressure_pu(rho).cpu().numpy()
+    print(p0.max(), p0.min(), p0.mean(), pfinal.max(), pfinal.min(), pfinal.mean())
+    print((p0-pfinal).max(), (p0-pfinal).min())
+    assert p0 == pytest.approx(pfinal, rel=0.0, abs=0.05)
```
That's the important part. I am initializing a TGV(2d and 3d). We know the initial pressure for both. I set the pressure to zero (rho = 1) and run jacobi.
2D works fine!!
3D does not converge to the correct value!!

Olllom commented on lettuce/util.py:
```diff
@@ -134,7 +138,53 @@ def torch_jacobi(f, p, dx, device, dim, tol_abs=1e-10):
+    # compute laplacian
+    u0 = torch_gradient(torch_gradient(u[0] * u[0], dx)[0] + torch_gradient(u[0] * u[1], dx)[1], dx)[0]  # uii+uij
+    u1 = torch_gradient(torch_gradient(u[1] * u[0], dx)[0] + torch_gradient(u[1] * u[1], dx)[1], dx)[1]  # uji+ujj
+    u_mod = (u0 + u1)
+    # TODO(@MCBs): I removed the minus sign before this (u0+u1) expression to make it work. No idea.
+    # There is a sign error somewhere, but I don't know if it's here
```
@McBs PLEASE check this
before it was `-(u0+u1)`, which is probably correct. I am too lazy to look up the equations.
If this should be a minus, you have a sign error somewhere in your jacobi scheme.

Olllom reviewed (commented):
See comments inline for explanations

Olllom commented on lettuce/util.py:
```diff
@@ -170,9 +169,17 @@ def pressure_poisson(units, u, rho0, tol_abs=1e-10, max_num_steps=100000):
+            for j in range(dim):
+                derivative = torch_gradient(torch_gradient(u[i]*u[j], dx)[i], dx)[j]
+                u_mod += derivative
+    # TODO(@MCBs): still not working in 3D
```
this doesn't help either
maybe I fucked it up though

Olllom replied:
it works in 2D though

PhiSpel committed:
Update CI.yml

try running CI on test-branch. I have issues with some import statements at https://github.com/PhiSpel/lettuce/tree/cuda-native-philipp
